students/johnpharmd/lesson06/mailroom4.py

```diff
@@ -0,0 +1,198 @@
+    ty_text = """
+                Dear {title} {last_name},
+                Thank you for your generous donation in the
+                amount of {donation} USD.
+              """
+    if __name__ == '__main__':
+        print(ty_text.format(**donor_dict))
+    else:
+        return ty_text.format(**donor_dict)
+
+
+def get_report():
+    print()
+    psv = ['Donor Name', '| Total Given', '| Num Gifts',
+           '| Average Gift']
+    if __name__ == '__main__':
+        print('{:<15}{:>12}{:>12}{:>12}'.format(psv[0], psv[1],
+              psv[2], psv[3]))
+        for i in range(55):
+            print('-', end='')
+        print()
+    new_list = [[donors_amts[donor]['donations'], donor,
+                donors_amts[donor]['num_of_donations']] for
+                donor in donors_amts]
+    new_list.sort(reverse=True)
+    for donor_list in new_list:
+        formatted_donor = ('{:<15}'.format(donor_list[1])
+                           + '{}{:>10}'.format(' $', donor_list[0])
+                           + '{:>12}'.format(donor_list[2])
+                           + '{}{:>11}'.format(' $',
+                           donor_list[0] // donor_list[2]))
+        if __name__ == '__main__':
+            print(formatted_donor)
+
+
+def send_letters():
+    d_a = donors_amts
+    title = ''
+    now = datetime.datetime.now()
+    for donor in d_a:
+        title = d_a[donor]['title'].strip('\'')
+        with open(donor + now.strftime('%Y%m%d')
+                  + '.txt', 'w') as of:
+            donor_dict = {'title': title,
+                          'last_name': donor,
+                          'donations': d_a[donor]['donations']}
+            letter_text = """
+            Dear {title} {last_name},
+            Thank you for your generous previous giving
+            in the amount of {donations} USD. Attached
+            is our most recent independent, third
+            party audit.
+
+            We hope you agree that we have been good
+            stewards of our donors' funds, and that
+            you will consider donating again to our
+            project.
+
+            Best wishes for continued success,
+            [Signature]
+            """
+            of.write(letter_text.format(**donor_dict))
+    if __name__ == '__main__':
+        print('Generated a letter, just now, for each donor in the db.\n')
+
+
+def program_run():
+    menu_dict = {'1': prepare_ty, '2': get_report,
+                 '3': send_letters, 'q': sys.exit}
+    main_text = '\n'.join((
+        'Choose from the following:',
+        '"1" - Send a "Thank You",',
+        '"2" - Create a Report,',
+        '"3" - Send Letters to All Donors, or',
+        '"q" to Quit: '
+      ))
+    while True:
+        print('\nMain Menu:')
+        response = input(main_text)
+        print()
+        try:
+            if response == 'q':
+                print('Program execution completed.')
+            menu_dict[response]()
```

Comment:
to keep the while loop cleaner you can use `partial` function:
```
from functools import partial
menu_dict = {'q': partial(sys.exit, "Program execution completed") ...}
# this will give a function object with input params  
```